opt: tests failing with fast_int_set_large build tag

Running `make test PKG='./pkg/sql/opt/xform' TAGS'=fast_int_set_large'` (or `./dev test pkg/sql/opt/xform -- --define gotags=bazel,gss,fast_int_set_large`) causes the error below. We're likely missing a `ColSet.Copy` call somewhere.

```
--- FAIL: TestRules (1.97s)
    test_log_scope.go:161: test logs captured to: /var/folders/3j/7c3w0jjd14g4n19lvvrxtm2r0000gp/T/logTestRules565569366
    test_log_scope.go:79: use -show-logs to present logs inline
    --- FAIL: TestRules/join (0.75s)
        opt_tester.go:618: /Users/marcus/go/src/github.com/cockroachdb/cockroach/pkg/sql/opt/xform/testdata/rules/join:11172: output cols mismatch: (1) vs (1,18)
            (1) assertion failure
            Wraps: (2) attached stack trace
              -- stack trace:
              | github.com/cockroachdb/cockroach/pkg/sql/opt/props.(*Relational).VerifyAgainst
              |         /Users/marcus/go/src/github.com/cockroachdb/cockroach/pkg/sql/opt/props/verify.go:87
              | github.com/cockroachdb/cockroach/pkg/sql/opt/memo.(*Memo).CheckExpr
              |         /Users/marcus/go/src/github.com/cockroachdb/cockroach/pkg/sql/opt/memo/check_expr.go:54                                                                                    | github.com/cockroachdb/cockroach/pkg/sql/opt/memo.(*Memo).AddLookupJoinToGroup
              |         /Users/marcus/go/src/github.com/cockroachdb/cockroach/pkg/sql/opt/memo/expr.og.go:24503
              | github.com/cockroachdb/cockroach/pkg/sql/opt/xform.(*explorer).exploreLookupJoin
              |         /Users/marcus/go/src/github.com/cockroachdb/cockroach/pkg/sql/opt/xform/explorer.og.go:2572                                                                                | github.com/cockroachdb/cockroach/pkg/sql/opt/xform.(*explorer).exploreGroupMember
              |         /Users/marcus/go/src/github.com/cockroachdb/cockroach/pkg/sql/opt/xform/explorer.og.go:38
              | github.com/cockroachdb/cockroach/pkg/sql/opt/xform.(*explorer).exploreGroup                                                                                                        |         /Users/marcus/go/src/github.com/cockroachdb/cockroach/pkg/sql/opt/xform/explorer.go:181
              | github.com/cockroachdb/cockroach/pkg/sql/opt/xform.(*Optimizer).optimizeGroup                                                                                                      |         /Users/marcus/go/src/github.com/cockroachdb/cockroach/pkg/sql/opt/xform/optimizer.go:528
              | github.com/cockroachdb/cockroach/pkg/sql/opt/xform.(*Optimizer).optimizeEnforcer
              |         /Users/marcus/go/src/github.com/cockroachdb/cockroach/pkg/sql/opt/xform/optimizer.go:685
              | github.com/cockroachdb/cockroach/pkg/sql/opt/xform.(*Optimizer).enforceProps                                                                                                       |         /Users/marcus/go/src/github.com/cockroachdb/cockroach/pkg/sql/opt/xform/optimizer.go:645
              | github.com/cockroachdb/cockroach/pkg/sql/opt/xform.(*Optimizer).optimizeGroupMember
              |         /Users/marcus/go/src/github.com/cockroachdb/cockroach/pkg/sql/opt/xform/optimizer.go:555
              | github.com/cockroachdb/cockroach/pkg/sql/opt/xform.(*Optimizer).optimizeGroup                                                                                                      |         /Users/marcus/go/src/github.com/cockroachdb/cockroach/pkg/sql/opt/xform/optimizer.go:515
              | github.com/cockroachdb/cockroach/pkg/sql/opt/xform.(*Optimizer).Optimize
              |         /Users/marcus/go/src/github.com/cockroachdb/cockroach/pkg/sql/opt/xform/optimizer.go:261
              | github.com/cockroachdb/cockroach/pkg/sql/opt/testutils/opttester.(*OptTester).optimizeExpr
              |         /Users/marcus/go/src/github.com/cockroachdb/cockroach/pkg/sql/opt/testutils/opttester/opt_tester.go:2236
              | github.com/cockroachdb/cockroach/pkg/sql/opt/testutils/opttester.(*OptTester).OptimizeWithTables
              |         /Users/marcus/go/src/github.com/cockroachdb/cockroach/pkg/sql/opt/testutils/opttester/opt_tester.go:1152
              | github.com/cockroachdb/cockroach/pkg/sql/opt/testutils/opttester.(*OptTester).Optimize
              |         /Users/marcus/go/src/github.com/cockroachdb/cockroach/pkg/sql/opt/testutils/opttester/opt_tester.go:1139
              | github.com/cockroachdb/cockroach/pkg/sql/opt/testutils/opttester.(*OptTester).RunCommand
              |         /Users/marcus/go/src/github.com/cockroachdb/cockroach/pkg/sql/opt/testutils/opttester/opt_tester.go:616
              | github.com/cockroachdb/cockroach/pkg/sql/opt/xform_test.runDataDrivenTest.func1.1
              |         /Users/marcus/go/src/github.com/cockroachdb/cockroach/pkg/sql/opt/xform/optimizer_test.go:258
              | github.com/cockroachdb/datadriven.runDirective.func1
              |         /Users/marcus/go/src/github.com/cockroachdb/cockroach/vendor/github.com/cockroachdb/datadriven/datadriven.go:321
              | github.com/cockroachdb/datadriven.runDirective
              |         /Users/marcus/go/src/github.com/cockroachdb/cockroach/vendor/github.com/cockroachdb/datadriven/datadriven.go:326
              | github.com/cockroachdb/datadriven.runDirectiveOrSubTest
              |         /Users/marcus/go/src/github.com/cockroachdb/cockroach/vendor/github.com/cockroachdb/datadriven/datadriven.go:195
              | github.com/cockroachdb/datadriven.runTestInternal
              |         /Users/marcus/go/src/github.com/cockroachdb/cockroach/vendor/github.com/cockroachdb/datadriven/datadriven.go:168
              | github.com/cockroachdb/datadriven.RunTest
              |         /Users/marcus/go/src/github.com/cockroachdb/cockroach/vendor/github.com/cockroachdb/datadriven/datadriven.go:136
              | github.com/cockroachdb/cockroach/pkg/sql/opt/xform_test.runDataDrivenTest.func1
              |         /Users/marcus/go/src/github.com/cockroachdb/cockroach/pkg/sql/opt/xform/optimizer_test.go:255
              | github.com/cockroachdb/datadriven.Walk
              |         /Users/marcus/go/src/github.com/cockroachdb/cockroach/vendor/github.com/cockroachdb/datadriven/datadriven.go:413
              | github.com/cockroachdb/datadriven.Walk.func1
              |         /Users/marcus/go/src/github.com/cockroachdb/cockroach/vendor/github.com/cockroachdb/datadriven/datadriven.go:426
              | testing.tRunner
              |         /usr/local/opt/go/libexec/src/testing/testing.go:1446
              | runtime.goexit
              |         /usr/local/opt/go/libexec/src/runtime/asm_amd64.s:1594
            Wraps: (3) output cols mismatch: (1) vs (1,18)
            Error types: (1) *assert.withAssertionFailure (2) *withstack.withStack (3) *errutil.leafError
    optimizer_test.go:202: -- test log scope end --
FAIL
FAIL    github.com/cockroachdb/cockroach/pkg/sql/opt/xform      4.728s
FAIL
```



Jira issue: CRDB-19702

--- a/pkg/sql/opt/xform/join_funcs.go
+++ b/pkg/sql/opt/xform/join_funcs.go
@@ -1237,6 +1237,9 @@ func (c *CustomFuncs) CreateLocalityOptimizedLookupJoinPrivateIncludingCols(
 	lookupExpr, remoteLookupExpr memo.FiltersExpr, private *memo.LookupJoinPrivate, cols opt.ColSet,
 ) *memo.LookupJoinPrivate {
 	newPrivate := c.CreateLocalityOptimizedLookupJoinPrivate(lookupExpr, remoteLookupExpr, private)
+	// Make a copy of the columns to avoid mutating the original
+	// LookupJoinPrivate's columns.
+	newPrivate.Cols = newPrivate.Cols.Copy()
 	newPrivate.Cols.UnionWith(cols)
 	return newPrivate
 }

